File: 1-twosum/solution.test.js
import { test, expect } from '@jest/globals'
import twoSum from './solution.js'

test('twoSum should work with positive numbers (& edge cases)', () => {
  expect(twoSum([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 19)).toEqual([8, 9])
  expect(twoSum([2, 7, 11, 15], 9)).toEqual([0, 1])
  expect(twoSum([3, 2, 4], 6)).toEqual([1, 2])
  expect(twoSum([3, 3], 6)).toEqual([0, 1])
	expect(twoSum([0, 4, 7, 3], 3)).toEqual([0, 3])

	// target is sum of first/last 2 elements
	expect(twoSum([4, 6, 8, 10], 10)).toEqual([0, 1])
  expect(twoSum([4, 6, 8, 10], 18)).toEqual([2, 3])
	expect(twoSum([2, 4, 7, 11], 6)).toEqual([0, 1])
	expect(twoSum([2, 4, 7, 11], 18)).toEqual([2, 3])
	expect(twoSum([0, 6, 8, 10], 6)).toEqual([0, 1])
	expect(twoSum([0, 4, 7, 11], 18)).toEqual([2, 3])

  expect(twoSum([5, 5, 5, 5], 10)).toEqual([0, 1]) // first pair (L2R)
})

test('twoSum should work with negative numbers', () => {
  expect(twoSum([-1, -2, -3, -4, -5], -8)).toEqual([2, 4])
	expect(twoSum([-1, -2, -3, -4, -5], -3)).toEqual([0, 1])
	expect(twoSum([-1, -2, -3, -4, -5], -9)).toEqual([3, 4])
	expect(twoSum([0, -2, -3, -4, -5], -5)).toEqual([1, 2])

	expect(twoSum([0, -2, -3, -4], -2)).toEqual([0, 1])
	expect(twoSum([1, -2, -3, -4], -7)).toEqual([2, 3])
	// should short circuit [0,1]:[1,-2] even if [2,3]:[3,-4] is valid
  expect(twoSum([1, -2, 3, -4], -1)).toEqual([0, 1])

	expect(twoSum([-5, -5, -5, -5], -10)).toEqual([0, 1])
})

test('twoSum should return `undefined` when no solution exists', () => {
  expect(twoSum([1, 2, 3], 7)).toEqual(undefined)
	expect(twoSum([-1, -2, -3], 0)).toEqual(undefined)
  expect(twoSum([1, 3, 5, 7], 2)).toEqual(undefined)
	expect(twoSum([-1, -3, -5, -7], 10)).toEqual(undefined)
})

test('twoSum should handle edge case of [{arr}.length = 1]', () => {
  expect(twoSum([1], 1)).toEqual(undefined)
	expect(twoSum([-1], -1)).toEqual(undefined)
})

test('twoSum should handle edge case of null(ish) input', () => {
	expect(twoSum(undefined, 1)).toEqual(undefined)
	expect(twoSum(null, 1)).toEqual(undefined)
	expect(twoSum([], 1)).toEqual(undefined)
})

File: 1-twosum/solution_test.go
package solutionTwoSum

import (
	"reflect"
	"testing"
)

type TestHarness struct {
	nums   []int
	target int
	expect []int
}

func checkHarness(t *testing.T, tests []TestHarness, suite string) {
	const err_strf = "TwoSum(%v, %d) = [got: %v], [expect: %v]"

	for _, tt := range tests {
		t.Run(suite, func(st *testing.T) {
			got := TwoSum(tt.nums, tt.target)
			if !reflect.DeepEqual(got, tt.expect) {
				st.Errorf(err_strf, tt.nums, tt.target, got, tt.expect)
			}
		})
	}
}

func TestTwoSum(t *testing.T) {
	basic_pos := []TestHarness{
		{nums: []int{1, 2, 3, 5, 6, 7, 8, 9}, target: 16, expect: []int{5, 7}},
		{nums: []int{2, 7, 11, 15}, target: 9, expect: []int{0, 1}},
		{nums: []int{3, 2, 4}, target: 6, expect: []int{1, 2}},
		{nums: []int{3, 3}, target: 6, expect: []int{0, 1}},
		{nums: []int{0, 4, 7, 3}, target: 3, expect: []int{0, 3}},

		// target is sum of first/last 2 elements
		{nums: []int{4, 6, 8, 10}, target: 10, expect: []int{0, 1}},
		{nums: []int{4, 6, 8, 10}, target: 18, expect: []int{2, 3}},
		{nums: []int{2, 4, 7, 11}, target: 6, expect: []int{0, 1}},
		{nums: []int{2, 4, 7, 11}, target: 18, expect: []int{2, 3}},
		{nums: []int{0, 6, 8, 10}, target: 6, expect: []int{0, 1}},
		{nums: []int{0, 4, 7, 11}, target: 18, expect: []int{2, 3}},

		// First pair (left to right)
		{nums: []int{5, 5, 5, 5}, target: 10, expect: []int{0, 1}},
	}
	basic_neg := []TestHarness{
		{nums: []int{-1, -2, -3, -4, -5}, target: -8, expect: []int{2, 4}},
		{nums: []int{-1, -2, -3, -4, -5}, target: -3, expect: []int{0, 1}},
		{nums: []int{-1, -2, -3, -4, -5}, target: -9, expect: []int{3, 4}},
		{nums: []int{0, -2, -3, -4, -5}, target: -5, expect: []int{1, 2}},
		{nums: []int{0, -2, -3, -4}, target: -2, expect: []int{0, 1}},
		{nums: []int{1, -2, -3, -4}, target: -7, expect: []int{2, 3}},
		// should short circuit [0,1]:[1,-2] even if [2,3]:[3,-4] is valid
		{nums: []int{1, -2, 3, -4}, target: -1, expect: []int{0, 1}},
		{nums: []int{-5, -5, -5, -5}, target: -10, expect: []int{0, 1}},
	}
	basic := append(basic_pos, basic_neg...)
	checkHarness(t, basic, "Basic positive/negative `nums`")

	edge_all := []TestHarness{
		{nums: []int{1, 2, 3}, target: 7, expect: nil},
		{nums: []int{-1, -2, -3}, target: 0, expect: nil},
		{nums: []int{1, 3, 5, 7}, target: 2, expect: nil},
		{nums: []int{-1, -3, -5, -7}, target: 10, expect: nil},

		{nums: []int{1}, target: 1, expect: nil},
		{nums: []int{-1}, target: -1, expect: nil},

		{nums: nil, target: 1, expect: nil},
		{nums: []int{}, target: 1, expect: nil},
	}
	checkHarness(t, edge_all, "No solution, single element, nil/empty input")
}

// Benchmark
func BenchmarkTwoSum(b *testing.B) {
	nums := []int{2, 7, 11, 15}
	target := 9
	for b.Loop() {
		TwoSum(nums, target)
	}

	lrg_nums := make([]int, 1000)
	for i := range lrg_nums {
		lrg_nums[i] = i
	}
	lrg_target := 1999
	for b.Loop() {
		TwoSum(lrg_nums, lrg_target)
	}
}
